tgui/packages/common/merge.js

```diff
@@ -0,0 +1,13 @@
+// Does a deep merge of two objects. DO NOT FEED CIRCULAR OBJECTS
+export const merge = (target, source) => {
+  return Object.entries(source).reduce((newArray, [key, value]) => {
+    newArray[key] =
+      value && typeof value === 'object'
+        ? merge(
+          (newArray[key] = newArray[key] || (Array.isArray(value) ? [] : {})),
+          value
+        )
+        : value;
+    return newArray;
+  }, target);
+};
```

Comment:
Replace with the following, and move it to `common/collections`, because it's working on objects which are technically collections. This function is immutable, so you don't need to clone objects beforehand.

```tsx
const isObject = (obj: unknown) => typeof obj === 'object' && obj !== null;

export const mergeDeep = (...objects: any[]): any => {
  const target = {};
  for (const object of objects) {
    for (const key of Object.keys(object)) {
      const targetValue = target[key];
      const objectValue = object[key];
      if (Array.isArray(targetValue) && Array.isArray(objectValue)) {
        target[key] = [...targetValue, ...objectValue];
      } else if (isObject(targetValue) && isObject(objectValue)) {
        target[key] = deepMerge(targetValue, objectValue);
      } else {
        target[key] = objectValue;
      }
    }
  }
  return target;
};
```

Reply:
Ah, much appreciated!